### DISCUSSION TITLE
Shelly 2.5 negative energy measurement on one channel

### DISCUSSION THREAD

**Discussion by kugelkopf123** [arendst/Tasmota#12943] [Q&A] [answered]
Hi all!

I've updated my Shellys 2.5 Shutter Relays to the latest branch and now one Channel shows negative values when switching on. 
The previous FW was 9.2.x and there were everything ok with the measurement. 
Has anyone an idea? 
 
<img width="441" alt="Bildschirmfoto 2021-08-22 um 21 03 08" src="https://user-images.githubusercontent.com/45996965/130366989-6d064aaa-4131-443d-bb71-9c5fcbf06c3a.png">
<img width="450" alt="Bildschirmfoto 2021-08-22 um 21 02 49" src="https://user-images.githubusercontent.com/45996965/130366992-8aed4d33-3232-4c7a-9ffe-91236ee75c5b.png">

**Comment by barbudor** [comment]
A minimum search would have you find this topic: https://github.com/arendst/Tasmota/discussions/12861

**Comment by kugelkopf123** [comment]
@barbudor Thanks for your feedback. I searched for it, but did not find this post. I suspected a problem with the Shelly2.5 because of the differences between the two channels. 
But your answer does not help me, because the linked topic only describes the possibility that negative values can now be displayed. However, a blind motor is connected to my Shelly2.5. So it does not produce a feed in. I havent found a command in the docs to reverse this value.

**Reply by barbudor** [reply]
My mistake, you are talking about Shelly 2.5 and the other article is about Shelly EM.

@pauloamexio your PR seems to have impacted Shelly 2.5
Can you please have a look?

**Reply by barbudor** [reply]
@kugelkopf123 
Looking at @pauloamexio, I see nothing wrong
Do you have a motor connect like on the 3rd diagram  of https://shelly.cloud/knowledge-base/devices/shelly-25/#wiring ?
Can you double-check your wiring ?
Not 100% sure sure but now that the sign is taken into account, you wiring may have an impact on how power is measured.

**Reply by kugelkopf123** [reply]
Yes, exactly like it is in the the third diagram. Ive check it. I'm switching the phase wires with the Shelly. Neutral is directly connected to the Motor. Just like in the Diagram.

**Reply by sfromis** [reply]
Checking, I also see negative numbers from Shelly 2.5 after going from 9.5.0.5 to 9.5.0.6 six days ago.

**Reply by pauloamexio** [reply]
Hello,
Please check the Shelly connection wires, because negative values are normal to happen when you measuring the wire inverted. 

Not sure, but perhaps if you are using motor you may need to change the O1 with the O2?

**Reply by sfromis** [reply]
I'm seeing negative numbers on a simple lamp (only using one channel), connections not touched for months, with Tasmota posting reasonable numbers till 6 days ago.

**Comment by arendst** [comment]
I have a shelly 2.5 too so will have a look into this.

**Comment by pauloamexio** [comment]
Hello,

I'm also have an shelly 2.5 in the light switch and what strange is that I have the channel A giving negative values, but the channel B is giving positive values:

![image](https://user-images.githubusercontent.com/87972768/130433933-64bb5b7f-f70a-494e-9194-26baa7fa580c.png)

I'm also need to check wiring connection to check if everything is ok.

**Comment by arendst** [comment]
No need to CHK wiring. Shelly internally swamped wires for channel A. I had that solved in the code. Latest change makes this fail.

I'll fix this asap

**Reply by pauloamexio** [reply]
Ok , and that part will not affect the Shelly em behaviour ?

**Reply by arendst** [reply]
Who knows ;-)

**Reply by miroslavpetrov** [reply]
@arendst Can you clarify what do you mean by "internally swamped wires"? The wires between the shunt resistors and the ade7953?

**Reply by arendst** [reply]
Well as the B channel (used by Relay1) inverts the power signal it seems the ADE7953 pins IBP and IBN have been swapped compared to the IAP and IAN pins. This results in negative power values as the current flows in the opposite way than on the IAP and IAN input pins which represent a positive power value.

**Reply by miroslavpetrov** [reply]
Thank you! I was experiencing such issues and was curious of the exact cause!

**Comment by arendst** [comment]
@pauloamexio for my understanding your Shaelly EM performs well? Both channel values are as expected?

If so I'll change the code to distinguish between different IRQ values. The default IRQ will support the shelly with channel A swapped; IRQ2 will support the EM with no channel swapped. So in your case you'll need another template.

I'll document once coding is done.

**Reply by pauloamexio** [reply]
Yes for now the shelly em is working as expect, but I have the Home general consumption/production connect in the channel 2:
 
![image](https://user-images.githubusercontent.com/87972768/130460689-c37645d8-0ced-445b-ba35-ed4e59a61c02.png)

**Comment by arendst** [comment]
OK. Just release a fix in the development branch.

For Shelly 2.5 nothing has changed. Same template as before can be used:
``{"NAME":"Shelly 2.5","GPIO":[320,0,32,0,224,193,0,0,640,192,608,225,3456,4736],"FLAG":0,"BASE":18}``

For Shelly EM a small change in the template is needed:
``{"NAME":"Shelly EM","GPIO":[0,0,0,0,0,0,0,0,640,3457,608,224,0,1],"FLAG":0,"BASE":18}``
Notice the increase from 3456 to 3457 to signal the code to support the Shelly EM instead of the Shelly 2.5

**Reply by pauloamexio** [reply]
The commit is this one:
https://github.com/arendst/Tasmota/commit/74b24298bde1603809fddd0f98b3956d38446985
right?

The way you made removes the feature of the shelly 2.5 to read negatives values, right? I know this kind of device is mostly use for lights or motors. But in cases where the shelly 2.5 is connect in some circuit that have inbound/outbound energy there will no way to identify the circuit direction :S 

For example if I want to replace my current shelly em for an shelly 2.5 I will had no way to understand if I was producing or consuming power...

**Reply by barbudor** [reply]
I think it doesn't make much sense to connect a Shelly 2.5 to a system that can reverse the power flow.
It is intended to drive a final equipment so any load balancing should be done up in the circuit.

**Reply by pauloamexio** [reply]
yes you're right ... it's possible that use case will never happen, but we're disable an feature from the device where the issue is only in one channel that the wire have been internally swamped

**Comment by arendst** [comment]
I'm looking into swapping channel B for the shelly 2.5 but it's an academic discussion as no-one ever asked for it.

That why I also requested what the Shelly EM does on relay 1. I don't know if that one is swapped too. Can you test this by using the other channel.

**Comment by arendst** [comment]
The latest comit allows for negative values on shelly 2.5 too.

As long as I don't know if the EM has no swapped hardware connection the above mentionent template differences are still true.

**Comment by pauloamexio** [comment]
For what I see in my board is that both splits are connect in the same direction:

![1629732228001](https://user-images.githubusercontent.com/87972768/130474483-825f4bcf-6809-4f1a-b1de-d6395da85be8.jpg)

So I don't think that the shelly em is also swapped.

**Comment by arendst** [comment]
OK. Thx. We'll keep the latest change than as solution. Still two templates needed.

**Comment by pauloamexio** [comment]
I did the update in the Shelly 2.5 and now the reactive power is negative... which is very strange:

![image](https://user-images.githubusercontent.com/87972768/130478921-5fad4520-80a1-4bc3-a722-8db0ebf1ea06.png)

The lamps that I'm using are leds

**Reply by pauloamexio** [reply]
the modifcations perharps must be only be done in the active_power and not in the reactive_power

**Comment by arendst** [comment, marked as answer]
Right. Reverted the last change. Back to excluding the Shelly 2.5 from negative numbers as was originally the case.

Expect the latest fix in compiled binary available over one hour.

**Reply by pauloamexio** [reply]
ok :)

**Reply by kugelkopf123** [reply]
Great! Now it works correct! Thanks for your help!

**Reply by sfromis** [reply]
My Shelly 2.5 devices also stopped being so negative...
